src/Features/Core/Portable/PullMemberUp/CodeActionAndSolutionGenerator.cs

```diff
@@ -0,0 +1,409 @@
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.CodeRefactorings.PullMemberUp;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+
+namespace Microsoft.CodeAnalysis.PullMemberUp
+{
+    internal class CodeActionAndSolutionGenerator
+    {
+        internal async Task<Solution> GetSolutionAsync(
+            AnalysisResult result,
+            Document contextDocument,
+            CancellationToken cancellationToken)
+        {
+            var codeGenerationService = contextDocument.Project.LanguageServices.GetRequiredService<ICodeGenerationService>();
+            var solution = contextDocument.Project.Solution;
+            var targetNodeSyntax = await codeGenerationService.FindMostRelevantNameSpaceOrTypeDeclarationAsync(solution, result.Target).ConfigureAwait(false);
+
+            if (targetNodeSyntax != null)
+            {
+                var solutionEditor = new SolutionEditor(solution);
+                var (editorMap, syntaxMap) = await InitializeEditorMapsAndSyntaxMap(result, targetNodeSyntax, solutionEditor, solution, cancellationToken).ConfigureAwait(false);
+                if (result.Target.TypeKind == TypeKind.Interface)
+                {
+                    return GenerateTargetingInterfaceSolution(
+                        result, codeGenerationService, targetNodeSyntax,
+                        solutionEditor, solution, editorMap, syntaxMap);
+                }
+                else
+                {
+                    return GenerateTargetingClassSolution(
+                        result, codeGenerationService, targetNodeSyntax,
+                        solutionEditor, solution, editorMap, syntaxMap);
+                }
+            }
+            else
+            {
+                return default;
+            }
+        }
+
+        internal CodeAction GetCodeActionAsync(
+            AnalysisResult result,
+            Document contextDocument,
+            string title)
+        {
+            if (result.Target.TypeKind == TypeKind.Interface)
+            {
+                var codeGenerationService = contextDocument.Project.LanguageServices.GetRequiredService<ICodeGenerationService>();
+                return CreateAddMembersToInterfaceAction(result, contextDocument, codeGenerationService, title);
+            }
+            else if (result.Target.TypeKind == TypeKind.Class)
+            {
+                return new SolutionChangeAction(
+                    title,
+                    async cancellationToken =>
+                    {
+                        var changedSolution = await GetSolutionAsync(result, contextDocument, cancellationToken).ConfigureAwait(false);
+                        return changedSolution;
+                    });
```

Comment:
nit.  lambda does not need to be async.  body does not need to use await/configureawait.